app/services/editor-commands/editor-commands.ts

```diff
@@ -0,0 +1,273 @@
+    undoMetadata: [],
+    redoMetadata: [],
+    operationInProgress: false,
+  };
+
+  // IMPORTANT: Must be kept in sync with metadata stored in
+  // this service's vuex state.
+  undoHistory: Command[] = [];
+  redoHistory: Command[] = [];
+
+  combineActive = false;
+
+  combineTimeout: number;
+
+  init() {
+    this.sceneCollectionsService.collectionWillSwitch.subscribe(() => this.clear());
+  }
+
+  private setCombineTimeout() {
+    this.combineActive = true;
+
+    if (this.combineTimeout) clearTimeout(this.combineTimeout);
+
+    this.combineTimeout = window.setTimeout(() => {
+      this.combineActive = false;
+      this.combineTimeout = null;
+    }, COMBINE_TIMEOUT);
+  }
+
+  executeCommand<TCommand extends keyof typeof COMMANDS>(
+    commandType: TCommand,
+    ...commandArgs: ConstructorParameters<(typeof COMMANDS)[TCommand]>
+  ) {
+    // Executing any command clears out the redo history, since we are
+    // creating a new branch in the timeline.
+    this.redoHistory = [];
+    this.CLEAR_REDO_METADATA();
+
+    const instance: Command = new (COMMANDS[commandType] as any)(...commandArgs);
+    const ret = instance.execute();
+
+    if (ret instanceof Promise) {
+      this.SET_OPERATION_IN_PROGRESS(true);
+      ret.then(() => this.SET_OPERATION_IN_PROGRESS(false));
+    }
+
+    if (instance instanceof CombinableCommand) {
+      if (this.combineActive) {
+        const previousCommand = this.undoHistory[this.undoHistory.length - 1];
+        if (previousCommand.constructor === instance.constructor) {
+          // This check is actually redundant since we just checked the constructor
+          // is the same above, but TypeScript isn't that smart.
+          if (previousCommand instanceof CombinableCommand) {
+            if (previousCommand.shouldCombine(instance)) {
+              previousCommand.combine(instance);
+              this.setCombineTimeout();
+              return;
+            }
+          }
+        }
+      }
+
+      this.setCombineTimeout();
+    }
+
+    this.undoHistory.push(instance);
+    this.PUSH_UNDO_METADATA({ description: instance.description });
+
+    return ret;
+  }
+
+  @shortcut('Ctrl+Z')
+  undo() {
+    if (this.state.operationInProgress) return;
+
+    const command = this.undoHistory.pop();
+    this.POP_UNDO_METADATA();
+
+    if (command) {
+      let ret: any;
+
+      try {
+        ret = command.rollback();
+      } catch (e) {
+        this.handleUndoRedoError(true, e);
+        return;
+      }
+
+      if (ret instanceof Promise) {
+        this.SET_OPERATION_IN_PROGRESS(true);
+        ret
+          .then(() => this.SET_OPERATION_IN_PROGRESS(false))
+          .catch(e => {
+            this.SET_OPERATION_IN_PROGRESS(false);
+            this.handleUndoRedoError(true, e);
+          });
+      }
+
+      this.redoHistory.push(command);
+      this.PUSH_REDO_METADATA({ description: command.description });
+    }
+  }
+
+  @shortcut('Ctrl+Y')
```

Comment:
We should add the `Ctrl+Y` hint somewhere in interface. It's not always obvious that `Ctrl+Y` is redo

Reply:
Hmm yeah.  I could make them accelerators in the right click menu instead, so they would show up there.  I could also add them to the icon tooltips.